### DISCUSSION TITLE
Quantum Ledger Driver for Mikro-ORM or Knex?

### DISCUSSION THREAD

**Discussion by jbmikk** [mikro-orm/mikro-orm#2287] [Q&A]
We have a system using Mikro-ORM with a Postgres database.

For a number of reasons (scalability and time travel, among others), we are considering moving to [Amazon's QLDB](https://aws.amazon.com/qldb/) (Quantum Ledger Database).

QLDB uses an SQL92 extension called [Partiql](https://partiql.org/) that is good for querying both traditional relational relational databases and embeded collections as well. So QLDB records look like documents, but they are queried using a syntax that looks pretty much like regular SQL.

The extension consists on treating nested collections as if they were separate tables you can join and filter. Like this:

```
SELECT 
       e.name AS employeeName, 
       p.name AS projectName
FROM 
     employees AS e, 
     e.projects AS p
WHERE p.name LIKE '%security%'
```

The only way I found of querying the database is using the standard [node drivers](https://github.com/awslabs/amazon-qldb-driver-nodejs).

So I'm thinking about writing a custom driver for QLDB so we don't have to rewrite the whole application without Mikro-ORM.

What we haven't decided yet is whether to use a custom knex dialect, or a custom Mikro-ORM driver, so I would like to hear some opinions about it.

Partiql being a standard SQL extensions makes me think a knex driver is a good idea, however those extensions seem so different I'm not sure knex's API would work well for making real life queries. On the other hand a Mikro-ORM driver seems flexible enough to handle something like this, there is Mongo support already, so it has to be possible.

**Comment by B4nan** [comment]
If you operate with the driver in a similar manner like mongo does (so not via SQL queries like knex), then it's better to base it on the mongo driver and ignore the SQL nature. 

Worth noting that mongo driver does not support joining and M:N relations are implemented differently (array of PK on the owning entity).
